File: ReSharper.FSharp/test/data/cache/csharpResolve/Members 02 - Virtual.fs
module Module

[<AbstractClass>]
type AbstractClass() =
    abstract member M1: unit -> unit
    default this.M1() = ()

    abstract member M2: int -> unit
    default this.M2(i: int) = ()

    abstract member M3<'P1>: int -> unit
    default this.M3<'P1>(i: int) = ()

    abstract member M4: 'P1 -> unit
    default this.M4(p: 'P1) = ()

    abstract member P1: int
    default this.P1 = 1

[<AbstractClass>]
type GenericAbstractClass<'TOuter>() =
    abstract member M1: unit -> unit
    default this.M1() = ()

    abstract member M2: int -> unit
    default this.M2(i: int) = ()

    abstract member M3<'P1>: int -> unit
    default this.M3<'P1>(i: int) = ()

    abstract member M4<'P1>: 'P1 -> unit
    default this.M4<'P1>(i: 'P1) = ()

    abstract member M5<'P1>: 'P1 * 'TOuter -> unit
    default this.M5<'P1>(i1: 'P1, i2: 'TOuter) = ()

    abstract member P1: 'TOuter
    default this.P1 = Unchecked.defaultof<'TOuter>

[<AbstractClass>]
type GenericAbstractClass2<'TSystem, 'TState, 'TInput, 'TOutput>() =
    abstract member M: state: 'TState * input: 'TInput -> bool
    default this.M(_, _) = true

type GenericClass<'TOuter>() =
    abstract member M: unit -> unit
    default this.M() = ()

    abstract member M2: int -> unit
    default this.M2(i: int) = ()
    
    abstract member M3<'P1>: int -> unit
    default this.M3<'P1>(i: int) = ()

    abstract member M4: 'TOuter -> unit
    default this.M4(p: 'TOuter) = ()
    
    abstract member M5<'P1>: 'P1 * 'TOuter -> unit
    default this.M5<'P1>(i1: 'P1, i2: 'TOuter) = ()

File: ReSharper.FSharp/test/data/cache/csharpResolve/Members 02 - Virtual.cs
public class InheritAbstractClass : Module.AbstractClass;

public class InheritAbstractClassOverrides : Module.AbstractClass
{
    public override void M1() => base.M1();
    public override void M2(int i) => base.M2(i);
    public override void M3<P1>(int i) => base.M3<P1>(i);
    public override void M4<P1>(P1 p1) => base.M4(p1);
    public override int P1 { get; }
}

public class InheritGenericAbstract : Module.GenericAbstractClass<int>;

public class InheritGenericAbstractOverrides : Module.GenericAbstractClass<int>
{
    public override void M1() => base.M1();
    public override void M2(int i) => base.M2(i);
    public override void M3<P1>(int i) => base.M3<P1>(i);
    public override void M4<P1>(P1 p1) => base.M4(p1);
    public override void M5<P1>(P1 p1, int i) => base.M5<P1>(p1, i);
    public override int P1 { get; }
}

public class InheritGeneric : Module.GenericClass<int>;

public class InheritGenericOverrides : Module.GenericClass<int>
{
    public override void M() => base.M();
    public override void M2(int i) => base.M2(i);
    public override void M3<P1>(int i) => base.M3<P1>(i);
    public override void M4(int i) => base.M4(i);
    public override void M5<P1>(P1 p1, int i) => base.M5<P1>(p1, i);
}

public sealed class Counter;
public sealed class CounterState;

public class InheritGenericCommand2 : Module.GenericAbstractClass2<Counter, CounterState, bool, int>;

public class InheritGenericCommand2Overrides : Module.GenericAbstractClass2<Counter, CounterState, bool, int>
{
    public override bool M(CounterState state, bool input) => base.M(state, input);
}
